pgxn/neon_walredo/shmpipe.c

```diff
@@ -0,0 +1,710 @@
+					resp_size = resp_hdr.size;
+					pipe->resp.tail.pos = (pipe->resp.tail.pos + sizeof resp_hdr) % QUEUE_BUF_SIZE;
+				}
+			}
+			else
+			{
+				assert(pipe->resp.busy);
+				if (available > resp_size)
+					available = resp_size;
+
+				if (tail >= available)
+				{
+					memcpy(resp, &pipe->resp.data[pipe->resp.tail.pos], available);
+				}
+				else
+				{
+					memcpy(resp, &pipe->resp.data[pipe->resp.tail.pos], tail);
+					memcpy(resp + tail, &pipe->resp.data[0], available - tail);
+				}
+				resp += available;
+				resp_size -= available;
+				pipe->resp.tail.pos = (pipe->resp.tail.pos + available) % QUEUE_BUF_SIZE;
+			}
+		}
+	}
+	/* Align position on header length */
+	pipe->resp.tail.pos = ALIGN(pipe->resp.tail.pos, MESSAGE_DATA_ALIGNMENT) % QUEUE_BUF_SIZE;
+
+#ifndef BUSY_WAIT_RESPONSES
+	if (pipe->resp.tail.n_blocked != 0)
+	{
+		pipe->resp.tail.n_blocked = 0;
+		event_signal(&pipe->resp.tail.event); /* Notify sender that tail is advanced */
+	}
+	if (pipe->resp.head.n_blocked != 0)
+	{
+		pipe->resp.head.n_blocked = 0;
+		event_signal(&pipe->resp.head.event); /* Wakeup other threads waiting for their responses */
+	}
+#endif
+
+	assert(pipe->resp.busy);
+	pipe->resp.busy = 0;
+
+	latch_release(&pipe->resp.cs);
+	return 1;
+}
+
+void shmem_pipe_get_request(pipe_t* pipe, char** data, u32* size, u32* msg_id)
+{
+	message_header_t req_hdr;
+	int header_received = 0;
+	size_t req_size = sizeof req_hdr;
+	char* req  = NULL;
+	size_t n_spin_iters = 0;
+
+	latch_acquire(&pipe->req.cs);
+
+	/* Take request from queue head */
+	while (req_size != 0)
+	{
+		size_t available = pipe->req.head.pos >= pipe->req.tail.pos
+			? pipe->req.head.pos - pipe->req.tail.pos
+			: QUEUE_BUF_SIZE - (pipe->req.tail.pos - pipe->req.head.pos);
+		if (available == 0)
+		{
+#ifndef BUSY_WAIT_RESPONSES
+			if (header_received && pipe->req.tail.n_blocked != 0)
+			{
+				pipe->req.tail.n_blocked = 0;
+				event_signal(&pipe->req.tail.event); /* Notify sender that tail is advanced */
+			}
+#endif
+			/* wait until head is advanced */
+			if (++n_spin_iters < MAX_SPIN_ITERATIONS)
+			{
+				/* Perform only limited number of busy loop iterations beause there may be no request from idle tenant for a long time */
+				latch_release(&pipe->req.cs);
+				RELEASE_CPU();
+				latch_acquire(&pipe->req.cs);
+			}
+			else
+			{
+				pipe->req.head.n_blocked += 1;
+				event_reset(&pipe->req.head.event);
+				latch_release(&pipe->req.cs);
+				event_wait(&pipe->req.head.event);
+				latch_acquire(&pipe->req.cs);
+			}
+		}
+		else
+		{
+			size_t tail = QUEUE_BUF_SIZE - pipe->req.tail.pos;
+			assert(available >= MESSAGE_DATA_ALIGNMENT);
+			if (!header_received)
+			{
+				assert(tail >= MESSAGE_DATA_ALIGNMENT);
+				memcpy(&req_hdr, &pipe->req.data[pipe->req.tail.pos], sizeof req_hdr);
+				header_received = 1;
+				req_size = req_hdr.size;
+				*msg_id = req_hdr.id;
+				*size = req_size;
+				req = *data = (char*)malloc(req_size);
```

Comment:
I think instead of malloc+free each request, it'd be better to only (re)allocate if `req_size > *size`, otherwise reuse the existing allocation, only free once.

Microbench results after the change are noise, but those requests are all the same size.

```patch
diff --git a/pgxn/neon_walredo/shmpipe.c b/pgxn/neon_walredo/shmpipe.c
index 62e239f6..b814b369 100644
--- a/pgxn/neon_walredo/shmpipe.c
+++ b/pgxn/neon_walredo/shmpipe.c
@@ -435,8 +435,22 @@ void shmem_pipe_get_request(pipe_t* pipe, char** data, u32* size, u32* msg_id)
                                header_received = 1;
                                req_size = req_hdr.size;
                                *msg_id = req_hdr.id;
+                               if (*size < req_size)
+                               {
+                                       if (*data == NULL)
+                                       {
+                                               req = *data = (char*)malloc(req_size);
+                                       }
+                                       else
+                                       {
+                                               req = *data = (char*)realloc(*data, req_size);
+                                       }
+                               } else {
+                                       // reuse existing allocation
+                                       req = *data;
+                               }
+                               assert(req != NULL);
                                *size = req_size;
-                               req = *data = (char*)malloc(req_size);
                                pipe->req.tail.pos = (pipe->req.tail.pos + sizeof req_hdr) % QUEUE_BUF_SIZE;
                        }
                        else
diff --git a/pgxn/neon_walredo/walredoproc.c b/pgxn/neon_walredo/walredoproc.c
index a465106e..a3496733 100644
--- a/pgxn/neon_walredo/walredoproc.c
+++ b/pgxn/neon_walredo/walredoproc.c
@@ -186,7 +186,7 @@ void
 WalRedoMain(int argc, char *argv[])
 {
        int                     firstchar;
-       StringInfoData input_message;
+       StringInfoData input_message = { 0 };
 #ifdef HAVE_LIBSECCOMP
        bool            enable_seccomp;
 #endif
@@ -379,8 +379,8 @@ WalRedoMain(int argc, char *argv[])
                        }
                        break;
                }
-               free(input_message.data);
        }                                                       /* end of input-reading loop */
+       free(input_message.data);
 }


```

Or maybe there's some reason to do it like this I don't understand?

Reply:
I do not think that one malloc per walredo request can have any measurable effect on performance.
If requests are small enough then  malloc() will most likely use block allocator which is very fast.
And if it is large ... then in any case you will need to do malloc.